solutions/cses-1091.mdx

```diff
@@ -51,3 +54,120 @@ int main(){
 	return 0;
 }
 ```
+
+</CPPSection>
+
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+
+public class ConcertTickets {
+	public static void main(String[] args) throws IOException {
+		Reader io = new Reader();
+		PrintWriter pw = new PrintWriter(System.out);
+
+		int ticketNum = io.nextInt();
+		int peopleNum = io.nextInt();
+		// No multiset in java so we'll have to use a TreeMap
+		NavigableMap<Integer, Integer> ticketMultiset = new TreeMap<>();
+		Map.Entry<Integer, Integer> val;
+
+		for (int i = 0; i < ticketNum; i++) {
+			int priceOfTicket = io.nextInt();
+			/*
+			 * If the set already contains the same key value, create another pair 
+			 * next to it
+			 */
+			if (ticketMultiset.containsKey(priceOfTicket)) {
+				ticketMultiset.put(priceOfTicket, ticketMultiset.get(priceOfTicket) + 1);
+			} else {
+				ticketMultiset.put(priceOfTicket, 1);
+			}
+		}
+
+		for (int i = 0; i < peopleNum; i++) {
+			int customerMaxPrice = io.nextInt();
+			/*
+			 * Get the optimal ticket for this customer 
+			 * We add 1 because lowerEntry returns the highest *strictly* lower entry
+			 */
+			val = ticketMultiset.lowerEntry(customerMaxPrice + 1); 
+			// If there are no possible prices, we can exit and return -1
+			if (val != null) {
+				pw.println(val.getKey());
+				/*
+				 * If there's more than one set with the key value, then replace
+				 * the current set with the next-lowest set with the same key 
+				 */
+				if (val.getValue() == 1) {
+					ticketMultiset.remove(val.getKey());
+
+				} else {
+					ticketMultiset.put(val.getKey(), val.getValue() - 1);
+				}
+			} else {
+				pw.println(-1);
+			}
+		}
+		io.close();
+		pw.close();
+	}
+	//BeginCodeSnip{Fast Reader}
```

Comment:
kattio? did you try that first?

Reply:
Yep, didn't work even in its most slimmed-down version.

Reply:
wow that's an f